model/src/main/kotlin/DependencyNavigator.kt

```diff
@@ -167,31 +167,44 @@ interface DependencyNavigator {
         allDependencies: Set<Identifier>
     ): Map<Identifier, List<Identifier>> {
         data class QueueItem(
-            val pkgRef: DependencyNode,
-            val parents: List<Identifier>
+            val node: DependencyNode,
+            val parent: DependencyNode?
         )
 
-        val remainingIds = allDependencies.toMutableSet()
+        // A node is visited if and only if it is a key in this map.
+        val predecessorForVisitedNode = mutableMapOf<DependencyNode, DependencyNode?>()
```

Comment:
Note to myself: This maps actually *does* need to store `null`-values as there is an iteration over the keys in line 195 below, which would behave differently if `null`-values were omitted from the map.